torch/fx/experimental/symbolic_shapes.py

```diff
@@ -683,6 +710,39 @@ def codegen_guards(self, placeholders, sources,
 
         from torch._dynamo.source import NegateSource, TensorPropertySource, TensorProperty
 
+        SourceInfo = collections.namedtuple("SourceInfo", ["base", "sources"])
+        SymbolicShapesExpr = collections.namedtuple("SymbolicShapeExpr", ["guard_expr", "call_expr", "fn"])
+        source_name_to_types = {}
+        source_name_to_infos = collections.defaultdict(lambda: SourceInfo(base=None, sources=[]))
+
+        def split_source_ref(source: Source) -> Tuple[str, str]:
+            ref = source_ref(source)
+            if isinstance(source, TensorPropertySource) and source.idx is not None:
+                end_position = -(1 + len(str(source.idx)) + 1)
+                return ref[0:end_position], ref[end_position:]
+            else:
+                return ref, ""
+
+        def flat_var_name(var_name: str) -> str:
+            var_name = var_name.replace(" ", "_")
```

Comment:
This seems pretty brittle.  Is it just for duplicate detection, perhaps a hash?

Reply:
I'm guessing you mean the first line of the `flat_var_name` function.

Reply:
yeah I meant the entire function.

Reply:
Ah, I see. I do agree it's brittle.
We were discussing about it [above](https://github.com/pytorch/pytorch/pull/89707#discussion_r1072998868).

**Context:** in order to JIT compile the symbolic shape expressions with Numba, we needed to extract them into their own function, i.e. `def ___symbolic_shape(...)`, where each parameter should correspond to each source.

**Problem:** the sources are Python expressions (e.g. `getattr(self.weights[0], '1')`). How to make them into Python identifiers?

**Solution:** I thought about 2 of them:
 
1. (current implementation) mangle the expression, replacing each character not allowed in an identifier.
    - Brittle, but we can map the mangled name back to the old name (good for debugging, I guess) 
2. assign a number to each unique source. The corresponding parameter will have name `var<N>`.
    - Better, but it makes inspection and debugging harder

Reply:
Side note, I think we want to do CSE on these expressions:
```
a = getattr(self.weights[0], '0')
b = getattr(self.weights[0], '1')
c = getattr(self.weights[0], '2')
```
should become
```
tmp = self.weights[0]
a = getattr(tmp, '0')
b = getattr(tmp, '1')
c = getattr(tmp, '2')
```

There are cases where this will make things asymptotically faster (guarding a linked list or a deeply nested `nn.Module` hierarchy), since current guards start from the root of the data structure every time.

Reply:
Got it. I think I will give it a go after moving away from Numba into C++.